server/sonar-ce/src/main/java/org/sonar/ce/container/CePluginRepository.java

```diff
@@ -0,0 +1,119 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.ce.container;
+
+import com.google.common.collect.ImmutableList;
+import java.io.File;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.commons.io.FileUtils;
+import org.picocontainer.Startable;
+import org.sonar.api.Plugin;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.core.platform.PluginInfo;
+import org.sonar.core.platform.PluginLoader;
+import org.sonar.core.platform.PluginRepository;
+import org.sonar.server.platform.DefaultServerFileSystem;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+
+/**
+ * Entry point to load plugins on startup. It assumes that plugins
+ * have been correctly installed/uninstalled/updated during web server startup
+ */
+public class CePluginRepository implements PluginRepository, Startable {
+
+  private static final String[] JAR_FILE_EXTENSIONS = new String[] {"jar"};
+
+  private final DefaultServerFileSystem fs;
+  private final PluginLoader loader;
+  private final AtomicBoolean started = new AtomicBoolean(false);
+
+  // following fields are available after startup
+  private final Map<String, PluginInfo> pluginInfosByKeys = new HashMap<>();
+  private final Map<String, Plugin> pluginInstancesByKeys = new HashMap<>();
+
+  public CePluginRepository(DefaultServerFileSystem fs, PluginLoader loader) {
+    this.fs = fs;
+    this.loader = loader;
+  }
+
+  @Override
+  public void start() {
+    Loggers.get(getClass()).info("Load plugins");
+    for (File file : listJarFiles(fs.getInstalledPluginsDir())) {
+      PluginInfo info = PluginInfo.create(file);
+      pluginInfosByKeys.put(info.getKey(), info);
+    }
+    pluginInstancesByKeys.putAll(loader.load(pluginInfosByKeys));
+    started.set(true);
+  }
+
+  @Override
+  public void stop() {
+    // close classloaders
+    loader.unload(pluginInstancesByKeys.values());
+    pluginInstancesByKeys.clear();
+    pluginInfosByKeys.clear();
+    started.set(false);
+  }
+
+  @Override
+  public Collection<PluginInfo> getPluginInfos() {
+    checkState(started.get(), "not started yet");
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "not started yet" 4 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1192)